connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/SparkResult.scala

```diff
@@ -198,6 +206,29 @@ private[sql] class SparkResult[T](
     nonEmpty
   }
 
+  private def processObservedMetrics(
+      metrics: java.util.List[ObservedMetrics]): Iterable[(String, Row)] = {
+    metrics.asScala.map { metric =>
+      assert(metric.getKeysCount == metric.getValuesCount)
+      var schema = new StructType()
+      val keys = mutable.ListBuffer.empty[String]
+      val values = mutable.ListBuffer.empty[Any]
+      (0 until metric.getKeysCount).map { i =>
+        val key = metric.getKeys(i)
+        val value = LiteralValueProtoConverter.toCatalystValue(metric.getValues(i))
+        schema = schema.add(key, LiteralValueProtoConverter.toDataType(value.getClass))
```

Comment:
The unconverted literal should have a `dataType`. Can we use that instead?

Reply:
I didn't see `.dataType` for the literal `metric.getValues(i)`, only `.getLiteralTypeCase`.
Also to get the actual DataType, we have to look at the values, for example for `CalendarInterval`. I think we can't avoid looking at the converted values anyway.